### ISSUE TITLE
Bitmap graphics on vic20

### ISSUE THREAD

**Issue by odflor** [spotlessmind1975/ugbasic#632]
Hi!

Why this program is not working on a VIC 20? Only appears a black screen.

>   BITMAP ENABLE
>   CLS BLACK
>   DO
>      PLOT RND(SCREEN WIDTH),RND(SCREEN HEIGHT),RND(SCREEN COLORS)
>   LOOP

I'm configure the VICE with all the memory and a NTSC machine.

I know that the VIC 20 have not a real "Graphic Mode", but normally is emulated by changing the character patterns.

![imagen](https://github.com/spotlessmind1975/ugbasic/assets/25032803/daa14961-80b1-4178-8ebd-49e670d8f41f)

**Comment by spotlessmind1975** [comment]
Hi @odflor , and **thank you for your issue report!**

From the very beginning of the ugBASIC implementation for the VIC-20, I was confronted with a very important limitation, which was precisely the absence of bitmap graphics. Or rather, if the **VIC-I** could reach 176×184 pixels (22 columns x 23 rows) in NTSC, while in PAL it could reach 224×256 pixels (equal to 28 columns x 32 rows). So it needed from a minimum of 506 characters to a maximum of 896 characters. Since each character set defines 256 characters, it was necessary to change the character map twice (NTSC) or even three times (PAL) at each vertical blank. However, by hooking up to the raster beam whose true position is not known and therefore it is necessary to carry out an initial calibration. If you go to this source you will see that there is still a lot of vestigial code that did the calibration on the `vic20`:
https://github.com/spotlessmind1975/ugbasic/blob/main/ugbc/src/hw/vic1/startup.asm#L142

In short: _it was a very tiring, slow and impractical approach_.

In the **4GRAVITY** game I, instead, used another approach: mapping the images drawn with `PUT IMAGE` on the characters, which will then be drawn. By doing this I was able to implement entire games that work in text mode. 

Take a look here: https://spotlessmind1975.itch.io/4gravity

The real oddity is that it's not working today, and I'm investigating.

**Comment by spotlessmind1975** [comment]
Hi @odflor , the problem has been fixed on COLDFIX v1.14.1 [rev. 20230824.2245].

**Comment by odflor** [comment]
Only for the records, I upgrade all my ugBASIC and this example ain't work.

No problem, I will not boring with graphic mode on VIC 20 for now.

**Comment by spotlessmind1975** [comment]
Hi @odflor, **first of all thank you for your feedback!**

Have you try this example? 

[What a Wonderful World](https://github.com/spotlessmind1975/ugbasic/blob/main/examples/contrib_what_a_wonderful_world.bas)

It uses graphics and it works on `vic20`.

Thank you again, and have a nice retrocoding!

**Comment by odflor** [comment]
Mr: spotlessmind1975
For Some reason, no PRG or D64 file created. Can't run project. Compiles well, no errors, but no binary file created. Try on c64 and VIC20.

Best Regards

    On Monday, September 11, 2023 at 06:29:40 PM GMT-4, Marco Spedaletti ***@***.***> wrote:  
 
 


Hi @odflor, first of all thank you for your feedback!

Have you try this example?

What a Wonderful World

It uses graphics and it works on vic20.

Thank you again, and have a nice retrocoding!

**Comment by spotlessmind1975** [comment]
Hi @odflor, first of all **thank you for your kindly feedback!** 

There was a problem in the Windows version of the compiler, which I completely missed. Now the compiler is able to produce an executable for that specific example. However, the example currently only works for `vic20`, for other reasons.

The issue has been fixed with COLDFIX v1.14.2 [rev. 20230912.1730].
